src/Domain/CampaignRepository.php

```diff
@@ -443,32 +444,126 @@ public function updateCampaignFromSFData(Campaign $campaign, array $campaignData
+            $qb->andWhere(
+                $qb->expr()->orX(
+                    $qb->expr()->like('charity.name', ':termForWhere'),
+                    $qb->expr()->like('c.name', ':termForWhere'),
+                    $whereSummaryMatches,
+                )
+            );
+            $qb->setParameter('termForWhere', $termWildcarded);
+        }
+    }
+
+    private function sortForSearch(
+        QueryBuilder $qb,
+        ?string $safeSortField,
+        string $sortDirection,
+        ?string $termWildcarded
+    ): void {
+        if ($safeSortField === 'relevance') {
+            $qb->addOrderBy(
+                <<<EOT
+            CASE
+                WHEN charity.name LIKE :termForOrder THEN 20
+                WHEN c.name LIKE LOWER(:termForOrder) THEN 10
+                WHEN LOWER(JSON_EXTRACT(c.salesforceData, '$.summary')) LIKE LOWER(:termForOrder) THEN 5
+                ELSE 0
+            END
+            EOT,
+                $sortDirection,
+            );
+            $qb->setParameter('termForOrder', $termWildcarded);
+        } elseif ($safeSortField !== null) {
+            $qb->addOrderBy($safeSortField, ($sortDirection === 'asc') ? 'asc' : 'desc');
+        }
+
+        $qb->addOrderBy('c.endDate', 'DESC');
+    }
+
     /**
      * @param 'asc'|'desc' $sortDirection
+     * @param array<string, string> $jsonMatchOneConditions Keyed on JSON key name. Value must exactly match the
+     *                                                      JSON property with the same key.
+     * @param array<string, string> $jsonMatchInListConditions Keyed on plural JSON key name. Value must exactly match
+     *                                                         one of the items in the JSON array with the same key.
      * @return list<Campaign>
      */
-    public function search(?string $sortField, string $sortDirection): array
-    {
+    public function search(
+        ?string $sortField,
+        string $sortDirection,
+        int $offset,
+        int $limit,
+        ?string $status,
+        array $jsonMatchOneConditions,
+        array $jsonMatchInListConditions,
+        ?string $term
+    ): array {
         $qb = $this->getEntityManager()->createQueryBuilder();
 
-        /** @var ?string $safeSortField */
         $safeSortField = match ($sortField) {
-            // @todo this field needs adding (maybe to a new CampaignStats table) for this to work.
-            // @todo when sorting by matchFundsRemaining works make it the default here. Add other allowable sort-by
-            //       fields to this match.
             'matchFundsRemaining' => throw new \Exception('Sorting by matchFundsRemaining not yet implemented'),
+            'relevance' => 'relevance',
             default => null,
         };
 
-        $qb = $qb->select('c')
+        if ($term === null && $safeSortField === 'relevance') {
+            throw new \Exception('Please provide a term to sort by relevance');
```

Comment:
This copy looks like its intended to be user-facing, but it's not an exception type that would be caught and shown to the user

Reply:
OK maybe just the politeness level made me think it was meant to be user facing, fine if this is dev facing copy and only to be thrown in case of a bug.